redis/src/sentinel.rs

```diff
@@ -0,0 +1,558 @@
+        && replica_info["port"].parse::<u16>().is_ok()
+}
+
+fn random_replica_index(max: usize) -> usize {
+    if max == 0 {
+        0
+    } else {
+        let nanos = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .subsec_nanos();
+        (nanos as usize) % max
+    }
+}
+
+fn try_connect_to_first_replica(
+    addresses: &Vec<(String, u16)>,
+    start_index: usize,
+) -> Result<Client, crate::RedisError> {
+    if addresses.is_empty() {
+        fail!((
+            ErrorKind::NoValidReplicasFound,
+            "No valid replica found in sentinel for given name",
+        ));
+    }
+
+    let mut last_err = None;
+    for i in 0..addresses.len() {
+        let index = (i + start_index) % addresses.len();
+        match Client::open(addresses[index].clone()) {
+            Ok(client) => return Ok(client),
+            Err(err) => last_err = Some(err),
+        }
+    }
+
+    // We can unwrap here because we know there is at least one address.
+    Err(last_err.expect("There should be at least one address"))
+}
+
+fn valid_addrs<'a>(
+    servers_info: Vec<HashMap<String, String>>,
+    validate: impl Fn(&HashMap<String, String>) -> bool + 'a,
+) -> impl Iterator<Item = (String, u16)> {
+    servers_info
+        .into_iter()
+        .filter(move |info| validate(info))
+        .map(|mut info| {
+            // We can unwrap here because we already checked everything
+            let ip = info.remove("ip").unwrap();
+            let port = info["port"].parse::<u16>().unwrap();
+            (ip, port)
+        })
+}
+
+fn check_role_result(result: &RedisResult<Vec<Value>>, target_role: &str) -> bool {
+    if let Ok(values) = result {
+        if !values.is_empty() {
+            if let Ok(role) = String::from_redis_value(&values[0]) {
+                if role.to_ascii_lowercase() == target_role {
+                    return true;
+                }
+            }
+        }
+    }
+    false
+}
+
+fn check_role(ip: &str, port: u16, target_role: &str) -> bool {
+    if let Ok(client) = Client::open((ip, port)) {
+        if let Ok(mut conn) = client.get_connection() {
+            let result: RedisResult<Vec<Value>> = crate::cmd("ROLE").query(&mut conn);
+            if check_role_result(&result, target_role) {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn get_first_valid_master(
+    masters: Vec<HashMap<String, String>>,
+    service_name: &str,
+) -> RedisResult<(String, u16)> {
+    for (ip, port) in valid_addrs(masters, |m| is_master_valid(m, service_name)) {
+        if check_role(ip.as_str(), port, "master") {
+            return Ok((ip, port));
+        }
+    }
+
+    fail!((
+        ErrorKind::MasterNameNotFound,
+        "Master with given name not found in sentinel",
+    ))
+}
+
+#[cfg(feature = "aio")]
+async fn async_check_role(ip: &str, port: u16, target_role: &str) -> bool {
+    if let Ok(client) = Client::open((ip, port)) {
```

Comment:
I believe this will fail if the server uses TLS. Either ask the user whether to use TLS or not, or assume that if the sentinel uses TLS, then the server does, too.

Reply:
I think that's a good solution, but if you're looking to the simplest solution, it's just returning host and port.